pkg/scheduler/queue/queue.go

```diff
@@ -54,237 +55,353 @@ type Request interface{}
+
+const (
+	registerConnection querierOperationType = iota
+	unregisterConnection
+	notifyShutdown
+	forgetDisconnected
+)
+
+type enqueueRequest struct {
+	userID      string
+	req         Request
+	maxQueriers int
+	successFn   func()
+	processed   chan error
+}
+
 func NewRequestQueue(maxOutstandingPerTenant int, forgetDelay time.Duration, queueLength *prometheus.GaugeVec, discardedRequests *prometheus.CounterVec, enqueueDuration prometheus.Histogram) *RequestQueue {
 	q := &RequestQueue{
-		queues:                  newUserQueues(maxOutstandingPerTenant, forgetDelay),
+		maxOutstandingPerTenant: maxOutstandingPerTenant,
+		forgetDelay:             forgetDelay,
 		connectedQuerierWorkers: atomic.NewInt32(0),
 		queueLength:             queueLength,
 		discardedRequests:       discardedRequests,
 		enqueueDuration:         enqueueDuration,
 	}
 
-	q.cond = contextCond{Cond: sync.NewCond(&q.mtx)}
-	q.Service = services.NewTimerService(forgetCheckPeriod, nil, q.forgetDisconnectedQueriers, q.stopping).WithName("request queue")
+	q.Service = services.NewTimerService(forgetCheckPeriod, q.starting, q.forgetDisconnectedQueriers, q.stop).WithName("request queue")
 
 	return q
 }
 
-// EnqueueRequest puts the request into the queue. MaxQueries is user-specific value that specifies how many queriers can
-// this user use (zero or negative = all queriers). It is passed to each EnqueueRequest, because it can change
-// between calls.
-//
-// If request is successfully enqueued, successFn is called with the lock held, before any querier can receive the request.
-func (q *RequestQueue) EnqueueRequest(userID string, req Request, maxQueriers int, successFn func()) error {
-	start := time.Now()
-	defer func() {
-		q.enqueueDuration.Observe(time.Since(start).Seconds())
-	}()
+func (q *RequestQueue) starting(_ context.Context) error {
+	q.stopRequested = make(chan struct{})
+	q.stopCompleted = make(chan struct{})
 
-	q.mtx.Lock()
-	defer q.mtx.Unlock()
+	// These channels must not be buffered so that we can detect when dispatcherLoop() has finished.
+	q.querierOperations = make(chan querierOperation)
+	q.enqueueRequests = make(chan enqueueRequest)
+	q.availableQueriers = make(chan *availableQuerier)
+	q.cancelledQueriers = make(chan *availableQuerier)
 
-	if q.stopped {
-		return ErrStopped
+	go q.dispatcherLoop()
+
+	return nil
+}
+
+func (q *RequestQueue) dispatcherLoop() {
+	stopping := false
+	queues := newUserQueues(q.maxOutstandingPerTenant, q.forgetDelay)
+	waitingQueriers := list.New()
+
+	for {
+		needToDispatchQueries := false
+
+		select {
+		case <-q.stopRequested:
+			// Nothing much to do here - fall through to the stop logic below to see if we can stop immediately.
+			stopping = true
+		case qe := <-q.querierOperations:
+			switch qe.operation {
+			case registerConnection:
+				q.connectedQuerierWorkers.Inc()
+				queues.addQuerierConnection(qe.querierID)
+				needToDispatchQueries = true // TODO: only do this if resharding actually happened
```

Comment:
[nit] I would add a comment to explain why there may be queries to dispatch. We could have a generic comment above the `switch qe.operation {` to explain for each of these operations we need to set `needToDispatchQueries = true`.

Reply:
Good point, added in ab047c8